### PR TITLE
fix: prevent Cursor startup hangs and Run control-plane deadlocks

### PR DESCRIPTION
**Pull request by vijankush**
## What was going wrong?

In OpenCode, Cursor models sometimes looked like they were **thinking forever** and never answered.

There were **two separate bugs**:

### Bug 1: Startup could wait forever
When OpenCode asks Cursor things like:
- “what models do I have?”
- “which server should I talk to?”
- “please refresh my login token”

…the code waited for the network response **with no time limit**.

If Cursor’s servers accepted the request but never finished sending the body, OpenCode just sat there. No error. No answer. Just spinning.

### Bug 2: A normal Cursor “check-in” could stall the chat
Once a chat/run started, Cursor often sends small **control messages**, like:
- “give me request context”
- “what’s the MCP state?”
- “please store/read this tiny KV blob”

These are **not real tools**. They’re housekeeping. Cursor expects an immediate reply.

The old code treated almost every one of those messages as:
> “Something important happened. Do **not** safely retry/recover this run.”

So if the connection hiccuped after those check-ins, OpenCode could get stuck: it owed Cursor replies, recovery looked “unsafe,” and the turn hung until a long timeout (~5 minutes) finally failed.

### 1. Add short timeouts to startup/login network calls
If those startup requests don’t finish in about **5 seconds**, fail with a clear error instead of spinning forever.

Important detail: we time out the **whole operation**, including reading the response body. Just canceling `fetch()` is not enough, because a stuck `json()` / `text()` read can ignore cancel.

### 2. Make run recovery smarter about “safe” vs “unsafe”
We did **not** invent retry/replay from scratch. OpenCode/Cursor recovery already existed.

This PR teaches the run loop:
- **Safe housekeeping** (`request_context`, `mcp_state`, normal KV get/set): answer them, and still allow recovery if needed
- **Real work** (visible text, real tool calls, messy/unknown must-reply frames): do **not** blindly replay

Also: if Cursor asks for a reply we truly cannot understand, **fail immediately** instead of hanging.

### 3. Tiny cleanup
Shared one small helper (`withAbortDeadline`) so connect/auth/version code use the same timeout race logic. No behavior change there.

## In one sentence
**Stop infinite waiting on startup calls, and stop treating Cursor’s normal housekeeping messages like dangerous side effects that block recovery.**

## How we tested

Automated:
- [x] `bun test` (553 passing)
- [x] typecheck
- [x] build

Live validation in OpenCode with the Cursor provider (all PASS):

| Probe | Model | Result |
|---|---|---|
| Text | `composer-2.5` | `CURSOR_TEXT_OK` |
| Stream | `composer-2.5` | `1..40` + `STREAM_OK` |
| Reasoning | `grok-4.5` high | `larger=37, smaller=15, GROK_HIGH_OK` |
| GPT High | `gpt-5.6-terra` high | `99242 TERRA_HIGH_OK` |
| 1M | `gpt-5.6-terra-1m` high | `TERRA_1M_OK` |
| Claude | `claude-opus-4-8` high | `no OPUS_OK` |
| Tool use | `composer-2.5` | README read → `TOOL_OK` |
| After cancel | session continue | `CURSOR_AFTER_CANCEL_OK` |
| Gemini | `gemini-3.5-flash` | `GEMINI_OK` |
| Sonnet | `claude-sonnet-5` high | `SONNET_OK` |
| Sol | `gpt-5.6-sol` high | `SOL_OK` |
| Multi-turn | `composer-2.5` | recalled codeword |
| Kimi | `kimi-k2.7-code` | `KIMI_OK` |

### CODE REVIEW

File: src/transport/connect.ts [hunk 1]
```diff
@@ -16,6 +16,49 @@ import {
...
+      `${operation} timeout must be a positive integer no greater than ${MAX_TIMER_MS}`,
+      { code: "CURSOR_INVALID_TIMEOUT" },
+    )
+  }
+  return timeoutMs
+}
+
+/**
+ * Bound a complete unary operation, not only fetch(). Response body readers can
+ * ignore abort signals, so Promise.race remains the authoritative deadline.
```
Comment (vijankush) [inline src/transport/connect.ts]: **Simple version: why this timeout helper exists**

Just canceling a network request is not enough.

Sometimes Cursor answers the “headers” part, then never finishes sending the body. In that case, waiting on `response.json()` / `response.text()` can hang forever.

So we race the whole operation against a timer. If it takes too long, we fail with a clear timeout error instead of spinning.

File: src/replay-safety.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+export type ReplayBarrierReason =
+  | "visible-text"
+  | "visible-reasoning"
+  | "display-tool-lifecycle"
+  | "non-control-exec"
+  | "stateful-interaction"
+  | "unknown-or-malformed-frame"
+
+export class AttemptReplaySafety {
```
Comment (vijankush) [inline src/replay-safety.ts:13]: **Simple version: why this file exists**

We are not adding a brand-new replay system.

Old behavior was too blunt:
- almost every Cursor message made the run look “unsafe to recover”

But Cursor often sends normal housekeeping messages (“got context?”, “MCP state?”, “save/read this tiny blob”). Those should be answered, and they should not by themselves block recovery.

This file is the “is this message dangerous or just housekeeping?” checker.

File: src/replay-safety.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    || !!update?.tool_call_completed
+    || !!decoded.exec
+    || !!decoded.kv
+    || !!decoded.execControl
+    || !!decoded.interactionQuery
+    || !!decoded.checkpointBytes?.length
+}
+
+/** Classify one decoded server frame without performing any protocol side effects. */
+export function analyzeReplayFrame(
```
Comment (vijankush) [inline src/replay-safety.ts:140]: **Simple version: this function only judges, it does not act**

`analyzeReplayFrame` looks at a message and answers:
- was this real progress?
- should this block safe recovery?

It does **not** send replies itself. The run loop (`pump`) still decides whether to reply, continue, or fail.

File: src/language-model.ts [hunk 4]
```diff
@@ -96,6 +98,34 @@ const DEFAULT_RETRY_POLICY = {
 } as const
 const MAX_RETRY_ATTEMPTS = 10
 const MAX_RETRY_DELAY_MS = 30_000
+const RUN_REQUEST_DECODE_FAILED = "CURSOR_RUN_REQUEST_DECODE_FAILED"
```
Comment (vijankush) [inline src/language-model.ts:101]: **Simple version: how the run loop uses this**

For each Cursor message:
1. Ask: housekeeping or real work?
2. If Cursor expects a reply to housekeeping, send it.
3. Only treat real work / weird must-reply messages as “don’t casually recover.”
4. If we truly cannot understand a message that needs a reply, fail right away instead of hanging.

File: src/protocol/struct.ts [hunk 5]
```diff
@@ -86,6 +86,105 @@ export function readAllFields(b: Uint8Array): RawField[] {
...
+/**
+ * Strictly walk a protobuf message for security/replay decisions.
+ *
+ * Unlike `readAllFields`, this parser consumes the complete input, preserves
+ * uint64 varints as bigint, represents fixed-width fields, and rejects invalid
+ * tags, truncation, overflow, groups, and unsupported wire types. It deliberately
+ * returns `undefined` instead of a partial result: callers must fail closed when
+ * deciding whether replay is safe.
+ */
+export function readAllFieldsStrict(bytes: Uint8Array): StrictRawField[] | undefined {
```
Comment (vijankush) [inline src/protocol/struct.ts:135]: **Simple version: why “strict” parsing**

Normal decoding can be a bit forgiving.

For recovery decisions, we want the opposite: if the bytes look broken/ambiguous, do **not** pretend it was a safe housekeeping message.

Broken/unknown “please reply” messages should fail clearly, not hang quietly.

File: src/auth.ts [hunk 6]
```diff
@@ -82,49 +83,101 @@ export type TokenPair = {
   refreshToken: string
 }
 
+async function withAuthRequestDeadline<T>(
```
Comment (vijankush) [inline src/auth.ts]: **Simple version: login/token calls can hang too**

The “thinking forever” bug was not only model discovery.

API key exchange / token refresh can also get stuck while reading the response body. So they use the same kind of timeout: fail after a few seconds instead of waiting forever.

File: test/run-interruption.test.ts [hunk 7]
```diff
@@ -155,6 +233,113 @@ describe("interrupted Cursor Run handling", () => {
...
+    expect(writes).toHaveLength(2)
+    const getReply = decodeMessage<any>("AgentClientMessage", writes[0]!).kv_client_message
+    expect(getReply.id).toBe(15)
+    expect(getReply.get_blob_result.blob_data).toEqual(new TextEncoder().encode("inline blob"))
+    const setReply = decodeMessage<any>("AgentClientMessage", writes[1]!).kv_client_message
+    expect(setReply.id).toBe(16)
+    expect(setReply.set_blob_result).toBeDefined()
+  })
+
+  it("answers decoded request-context and MCP probes despite unknown fields", async () => {
```
Comment (vijankush) [inline test/run-interruption.test.ts:280]: **Simple version: these tests lock in the live bugfix**

They check that we:
- still answer Cursor’s housekeeping messages
- do not casually recover after ambiguous/broken ones
- fail fast instead of hanging when a required reply cannot be handled

If someone later brings back the old “every message is unsafe” behavior, these tests should catch it.

**Comment by vijankush** [comment]
## Small cleanup commit (no behavior change)

We noticed the same timeout race was copied in a few places, so we moved it into one helper: `withAbortDeadline` in `src/deadline.ts`.

Used by:
- model/server startup calls
- auth token exchange/refresh
- installer version fetch

Same behavior, less duplicated timeout code.

**Comment by vijankush** [comment]
## Plain-English summary

### Are we adding replay support?
**No.** Recovery/retry already existed.

### What did we fix?
The old code treated too many normal Cursor housekeeping messages as “unsafe,” which helped cause long “thinking forever” hangs.

### Mental model
1. **Housekeeping** (“got state?”, “save this blob?”) → answer it; recovery can still be okay
2. **Real work** (answer text, real tools) → don’t casually replay

### Also
Startup/login network calls now time out in about 5 seconds, including stuck response-body reads.

**One-liner:** stop infinite waits, and stop safe check-ins from blocking recovery.

**Comment by oakimov** [comment]
Reviewed in full — logic, error handling, and replay-safety semantics all check out. Traced the `failRunProtocol` → `pumpWithRecovery` path end-to-end: `CursorProtocolError.transient` is always `false`, so every protocol failure from this PR fails hard with no retry, independent of `AttemptReplaySafety`. That's correct by design (fail closed on protocol violations), not a gap.

I added two small test-coverage additions (no production code changes) to close gaps found while reviewing:

- `test/struct.test.ts`: `readAllFieldsStrict` didn't have coverage for field number `0` (must reject) or the `MAX_PROTOBUF_FIELD_NUMBER` boundary (`0x1fffffff` accepted, `0x20000000` rejected) — both are enforced in the code but untested.
- `test/client-version.test.ts`: `fetchInstallerVersion` had no test for its `withAbortDeadline` timeout path, unlike the analogous `AvailableModels`/`GetServerConfig`/auth calls in `connect-origin.test.ts`. Added one using a fetch mock that only resolves via the abort signal — verified it fails/hangs at the 5s mark if the `withAbortDeadline` wrapper is removed, and passes in ~20ms with it in place.

Pushed both onto a same-named `pr/cursor-startup-liveness` branch on this repo (branched from your `65d9311`) since I can't push to your fork directly: https://github.com/oakimov/cursor-opencode-provider/tree/pr/cursor-startup-liveness

@vijankush — feel free to pull/cherry-pick that commit (`46d474f` equivalent) into your branch if you'd like it in the PR itself; otherwise I'm happy to land it as a follow-up after merge. Either way, approving — full suite passes (553 → 556 with these additions), typecheck clean, no behavioral concerns.

**Review by oakimov — APPROVED** [review]
Approving. Full review covered all 14 changed files:

- `withAbortDeadline` correctly bounds the whole operation (fetch + body read), not just `fetch()` itself — closes the real gap `AbortSignal.timeout` leaves open.
- `readAllFieldsStrict` is a sound, fail-closed wire-level parser (rejects truncation/overflow/groups, preserves uint64 precision).
- `AttemptReplaySafety` / `analyzeReplayFrame` correctly reclassify housekeeping frames (request_context, mcp_state, KV) as replay-safe while keeping visible text/reasoning, tool lifecycle, and malformed frames as hard barriers.
- Traced `failRunProtocol` → `pumpWithRecovery`: `CursorProtocolError.transient` is always `false`, so these new protocol failures never retry regardless of `replaySafe` — correct fail-closed behavior, no unhandled-rejection risk.
- Test coverage is thorough (553 passing) and the new tests directly exercise the wire-level edge cases this PR is designed to protect.

Left two small test-coverage additions (see PR comment) on a `pr/cursor-startup-liveness` branch on this repo — no production code changes, purely closing untested edges I found while reviewing. Not blocking.

**Comment by vijankush** [comment]
@oakimov feel free to merge whenever!! Thank you :)
